### PR TITLE
refactor: Introduce Resource base class

### PR DESCRIPTION
**Pull request by dmaretskyi**
### Motivation / Background

This Pull Request ...

### CODE REVIEW

File: packages/common/util/src/di-key.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+// TODO(dmaretskyi): Collections.
+export type DiKey<T> = { new (...args: any[]): T } | SymbolDiKey<T>;
+
+export const DiKey = new (class DiKeyConstructor {
+  // TODO(dmaretskyi): Disable private members lowering for dev env.
+  // TODO(dmaretskyi): Could be a weak map after a NodeJS upgrade.
+  #combinedRegistry = new Map();
+
+  define<T>(name: string): DiKey<T> {
+    return Symbol(name) as any;
```
Comment (dmaretskyi) [inline packages/common/util/src/di-key.ts:39]: change to SymbolDiKey

File: packages/common/util/src/di-key.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      return value;
+    }
+  }
+
+  #combinedDescription([first, ...rest]: DiKey<any>[]) {
+    return `${this.stringify(first)}<${rest.map(this.stringify).join(', ')}>`;
+  }
+})();
+
+// class Di {
```
Comment (dmaretskyi) [inline packages/common/util/src/di-key.ts:82]: remove

File: packages/common/util/src/di-key.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  [inspect.custom]() {
+    return this.toString();
+  }
+}
+
+// TODO(dmaretskyi): Default values.
+// TODO(dmaretskyi): Collections.
+export type DiKey<T> = { new (...args: any[]): T } | SymbolDiKey<T>;
+
+export const DiKey = new (class DiKeyConstructor {
```
Comment (dmaretskyi) [inline packages/common/util/src/di-key.ts:33]: docs

Comment (dmaretskyi) [inline packages/common/util/src/di-key.ts:33]: docs

File: packages/common/context/src/resource.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+/**
+ * Base class for resources that need to be opened and closed.
+ */
+export class Resource implements Lifecycle {
+  #lifecycleState = LifecycleState.CLOSED;
+  #openPromise: Promise<void> | null = null;
+  #closePromise: Promise<void> | null = null;
+  #ctx: Context = this.#createContext();
```
Comment (yarolegovich) [inline packages/common/context/src/resource.ts:26]: not sure I understand what this context is for. also I think parentContext should be created using createContext?

File: packages/common/context/src/resource.ts [hunk 5]
```diff
@@ -0,0 +1,131 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import '@dxos/util';
+import { Context } from './context';
+
+export interface Lifecycle {
+  open?(ctx?: Context): Promise<any> | any;
```
Comment (yarolegovich) [inline packages/common/context/src/resource.ts:9]: why are methods optional? 🤔

Comment (dmaretskyi) [reply packages/common/context/src/resource.ts:9]: To allow typescript to structurally match types that aren't `Resource`. For example `openInContext` below could work on any time that has open and/or close.

File: packages/common/context/src/resource.ts (hunk 4 above)
Comment (dmaretskyi) [reply packages/common/context/src/resource.ts:31]: I've added jsdoc comments for them

**Pull request merged**